### PR TITLE
HID: Implement GC controller in game

### PR DESCRIPTION
**Pull request by german77**
Implements the necessary changes for the GC controller to be detected in-game as a GC controller. Analog triggers are going to be implemented in a future PR

### CODE REVIEW

File: src/yuzu/configuration/configure_input_player.cpp [hunk 1]
```diff
@@ -461,6 +465,7 @@ ConfigureInputPlayer::ConfigureInputPlayer(QWidget* parent, std::size_t player_i
         tr("Dual Joycons"),
         tr("Left Joycon"),
         tr("Right Joycon"),
+        tr("Game Cube"),
```
Comment (lioncash) [inline src/yuzu/configuration/configure_input_player.cpp:468]:
```suggestion
        tr("GameCube Controller"),
```

**Review by lioncash — COMMENTED** [review]
Localization-only review

File: src/input_common/settings.h [hunk 2]
```diff
@@ -339,6 +339,8 @@ enum class ControllerType {
     DualJoyconDetached,
     LeftJoycon,
     RightJoycon,
+    GameCube,
+    Pokeball,
     Handheld,
```
Comment (Morph1984) [inline src/input_common/settings.h:344]:
```suggestion
    LeftJoycon,
    RightJoycon,
    Handheld,
    GameCube,
    Pokeball,
```

These should be placed below handheld to preserve compatibility

File: src/yuzu/applets/controller.cpp [hunk 3]
```diff
@@ -85,6 +87,8 @@ constexpr Settings::ControllerType GetControllerTypeFromIndex(int index) {
     case 3:
         return Settings::ControllerType::RightJoycon;
     case 4:
+        return Settings::ControllerType::GameCube;
```
Comment (Morph1984) [inline src/yuzu/applets/controller.cpp:90]: ditto

File: src/yuzu/applets/controller.cpp [hunk 4]
```diff
@@ -101,8 +105,10 @@ constexpr int GetIndexFromControllerType(Settings::ControllerType type) {
         return 2;
     case Settings::ControllerType::RightJoycon:
         return 3;
-    case Settings::ControllerType::Handheld:
+    case Settings::ControllerType::GameCube:
         return 4;
+    case Settings::ControllerType::Handheld:
+        return 5;
```
Comment (Morph1984) [inline src/yuzu/applets/controller.cpp:111]: ditto

File: src/yuzu/configuration/configure_input_player.cpp [hunk 5]
```diff
@@ -68,6 +68,8 @@ constexpr Settings::ControllerType GetControllerTypeFromIndex(int index) {
     case 3:
         return Settings::ControllerType::RightJoycon;
     case 4:
+        return Settings::ControllerType::GameCube;
```
Comment (Morph1984) [inline src/yuzu/configuration/configure_input_player.cpp:71]: ditto

File: src/yuzu/configuration/configure_input_player.cpp [hunk 6]
```diff
@@ -84,8 +86,10 @@ constexpr int GetIndexFromControllerType(Settings::ControllerType type) {
         return 2;
     case Settings::ControllerType::RightJoycon:
         return 3;
-    case Settings::ControllerType::Handheld:
+    case Settings::ControllerType::GameCube:
         return 4;
+    case Settings::ControllerType::Handheld:
+        return 5;
```
Comment (Morph1984) [inline src/yuzu/configuration/configure_input_player.cpp:92]: ditto

File: src/yuzu/configuration/configure_input_player.cpp [hunk 7]
```diff
@@ -900,6 +907,12 @@ void ConfigureInputPlayer::UpdateControllerAvailableButtons() {
         ui->buttonShoulderButtonsRight,
         ui->buttonMiscButtonsPlusHome,
         ui->bottomRight,
+        ui->buttonMiscButtonsMinusGroup,
+        ui->buttonMiscButtonsScreenshotGroup,
+        ui->buttonMiscButtonsHomeGroup,
+        ui->buttonLStickPressedGroup,
+        ui->groupRStickPressed,
+        ui->buttonShoulderButtonsButtonZLGroup,
```
Comment (Morph1984) [inline src/yuzu/configuration/configure_input_player.cpp:915]: These should be disabled, not hidden

**Review by Morph1984 — COMMENTED** [review]
Review comments in addition to the ones discussed on discord

File: src/core/hle/service/hid/controllers/npad.cpp [hunk 8]
```diff
@@ -423,10 +461,14 @@ void Controller_NPad::OnUpdate(const Core::Timing::CoreTiming& core_timing, u8*
         auto& pokeball_entry =
             npad.pokeball_states.npad[npad.pokeball_states.common.last_entry_index];
         auto& libnx_entry = npad.libnx.npad[npad.libnx.common.last_entry_index];
+        auto& trigger_entry =
+            npad.gc_trigger_states.trigger[npad.gc_trigger_states.common.last_entry_index];
 
         libnx_entry.connection_status.raw = 0;
         libnx_entry.connection_status.IsConnected.Assign(1);
 
+        LOG_ERROR(Service_HID, "called, npad_id_1={}",
```
Comment (Morph1984) [inline src/core/hle/service/hid/controllers/npad.cpp:470]: Is this left over from debugging?

File: src/core/hle/service/hid/controllers/npad.cpp [hunk 9]
```diff
@@ -630,11 +685,14 @@ void Controller_NPad::OnMotionUpdate(const Core::Timing::CoreTiming& core_timing
             }
             break;
         case NPadControllerType::Pokeball:
+        case NPadControllerType::GameCube:
```
Comment (Morph1984) [inline src/core/hle/service/hid/controllers/npad.cpp:688]:
```suggestion
        case NPadControllerType::GameCube:
        case NPadControllerType::Pokeball:
```

**Review by slashiee — CHANGES_REQUESTED** [review]
There are duplicate icons appended with ` - copia`.

**Comment by german77** [comment]
> There are duplicate icons appended with ` - copia`.

I just separated the changes for this PR from my build. In a few hours I will properly remove all errors and test the build.  Please consider checking it tomorrow

File: src/core/hle/service/hid/controllers/npad.cpp [hunk 10]
```diff
@@ -175,6 +184,13 @@ void Controller_NPad::InitNewlyAddedController(std::size_t controller_idx) {
         controller.properties.use_plus.Assign(1);
         controller.pad_assignment = NpadAssignments::Single;
         break;
+    case NPadControllerType::GameCube:
+        controller.joy_styles.gamecube.Assign(1);
+        // GC controllers behave like pro controllers
```
Comment (Morph1984) [inline src/core/hle/service/hid/controllers/npad.cpp:189]:
```suggestion
        // The GC Controller behaves like a wired Pro Controller
```

File: src/core/hle/service/hid/controllers/npad.h [hunk 11]
```diff
@@ -104,7 +106,11 @@ class Controller_NPad final : public ControllerBase {
             BitField<3, 1, u32> joycon_left;
             BitField<4, 1, u32> joycon_right;
 
-            BitField<6, 1, u32> pokeball; // TODO(ogniK): Confirm when possible
+            BitField<5, 1, u32> gamecube;
+            BitField<6, 1, u32> pokeball;     // TODO(ogniK): Confirm when possible
+            BitField<7, 1, u32> lark;         // TODO: Confirm when possible
+            BitField<8, 1, u32> handheldlark; // TODO: Confirm when possible
+            BitField<9, 1, u32> lucia;        // TODO: Confirm when possible
```
Comment (Morph1984) [inline src/core/hle/service/hid/controllers/npad.h:113]: These are confirmed by switchbrew, so they can be removed

File: src/yuzu/configuration/configure_input_player.cpp [hunk 12]
```diff
@@ -961,6 +1048,11 @@ void ConfigureInputPlayer::UpdateMotionButtons() {
         ui->buttonMotionLeftGroup->hide();
         ui->buttonMotionRightGroup->show();
         break;
+    case Settings::ControllerType::GameCube:
+        // hide "Motion 1" and hide "Motion 2".
```
Comment (Morph1984) [inline src/yuzu/configuration/configure_input_player.cpp:1052]:
```suggestion
        // Hide both "Motion 1/2".
```

File: src/yuzu/configuration/configure_input_player.h [hunk 13]
```diff
@@ -121,6 +121,9 @@ class ConfigureInputPlayer : public QWidget {
     /// Hides and disables controller settings based on the current controller type.
     void UpdateControllerAvailableButtons();
 
+    /// Disables controller settings based on the current controller type.
+    void UpdateControllerEnableButtons();
```
Comment (Morph1984) [inline src/yuzu/configuration/configure_input_player.h:125]:
```suggestion
    void UpdateControllerEnabledButtons();
```

File: src/yuzu/applets/controller.cpp [hunk 14]
```diff
@@ -460,6 +467,7 @@ void QtControllerSelectorDialog::UpdateControllerIcon(std::size_t player_index)
     const QString stylesheet = [this, player_index] {
         switch (GetControllerTypeFromIndex(emulated_controllers[player_index]->currentIndex(),
                                            player_index)) {
+        case Settings::ControllerType::GameCube:
         case Settings::ControllerType::ProController:
```
Comment (Morph1984) [inline src/yuzu/applets/controller.cpp:471]: minor nit, swap the 2
```suggestion
        case Settings::ControllerType::ProController:
        case Settings::ControllerType::GameCube:
```

File: src/yuzu/configuration/configure_input_player.cpp [hunk 15]
```diff
@@ -882,6 +946,7 @@ void ConfigureInputPlayer::UpdateControllerIcon() {
     // "nonstandard" to use an image through the icon support)
     const QString stylesheet = [this] {
         switch (GetControllerTypeFromIndex(ui->comboControllerType->currentIndex())) {
+        case Settings::ControllerType::Pokeball:
         case Settings::ControllerType::ProController:
```
Comment (Morph1984) [inline src/yuzu/configuration/configure_input_player.cpp:950]: minor nit, swap the 2
```suggestion
        case Settings::ControllerType::ProController:
        case Settings::ControllerType::Pokeball:
```

File: src/yuzu/configuration/configure_input_player.h [hunk 16]
```diff
@@ -131,9 +131,15 @@ class ConfigureInputPlayer : public QWidget {
     /// Hides and disables controller settings based on the current controller type.
     void UpdateControllerAvailableButtons();
 
+    /// Disables controller settings based on the current controller type.
+    void UpdateControllerEnabledButtons();
+
     /// Shows or hides motion groupboxes based on the current controller type.
     void UpdateMotionButtons();
 
+    /// Alters the button name based on the current controller type.
```
Comment (Morph1984) [inline src/yuzu/configuration/configure_input_player.h:140]:
```suggestion
    /// Alters the button names based on the current controller type.
```

**Comment by v1993** [comment]
I've tracked down few bizarre issues to this PR and have to ask for its exclusion from EA merge.

Currently, there are two issues known to me:

1. Incorrect handling of ZL/ZR triggers, tested for Pro Controller input. Assigning buttons still interprets them as axles, causing weird results where assigned buttons don't work but unrelated axis does! This also breaks "default" configuration.
2. Right stick does not show informational prompt when being configured.

File: src/input_common/sdl/sdl_impl.cpp [hunk 17]
```diff
@@ -390,6 +390,64 @@ class SDLAnalog final : public Input::AnalogDevice {
...
+};
+
+class SDLDirectionTrigger final : public Input::AnalogTriggerDevice {
+public:
+    explicit SDLDirectionTrigger(std::shared_ptr<SDLJoystick> joystick_, int hat_, Uint8 direction_)
+        : joystick(std::move(joystick_)), hat(hat_), direction(direction_) {}
+
+    std::tuple<bool, float> GetStatus() const override {
+        if (joystick->GetHatDirection(hat, direction)) {
+            return {true, 1.0f};
```
Comment (ameerj) [inline src/input_common/sdl/sdl_impl.cpp:423]: Is the bool associated with a button?

File: src/input_common/sdl/sdl_impl.cpp [hunk 18]
```diff
@@ -584,6 +642,61 @@ class SDLAnalogFactory final : public Input::Factory<Input::AnalogDevice> {
...
+        const Common::ParamPackage& params) override {
+        const std::string guid = params.Get("guid", "0");
+        const int port = params.Get("port", 0);
+        auto joystick = state.GetSDLJoystickByGUID(guid, port);
+
+        if (params.Has("hat")) {
+            const int hat = params.Get("hat", 0);
+            const std::string direction_name = params.Get("direction", "");
+            Uint8 direction;
+            if (direction_name == "up") {
```
Comment (ameerj) [inline src/input_common/sdl/sdl_impl.cpp:666]: maybe use a switch statement?

File: src/input_common/sdl/sdl_impl.cpp [hunk 19]
```diff
@@ -584,6 +642,61 @@ class SDLAnalogFactory final : public Input::Factory<Input::AnalogDevice> {
...
+            }
+            // This is necessary so accessing GetHat with hat won't crash
+            joystick->SetHat(hat, SDL_HAT_CENTERED);
+            return std::make_unique<SDLDirectionTrigger>(joystick, hat, direction);
+        }
+        if (params.Has("axis")) {
+            const int axis = params.Get("axis", 0);
+            const float deadzone = std::clamp(params.Get("deadzone", 0.0f), 0.0f, 1.0f);
+            const float range = std::clamp(params.Get("range", 1.0f), 0.50f, 1.50f);
+            const float threshold = std::clamp(params.Get("threshold", 0.9f), 0.0f, 1.0f);
```
Comment (ameerj) [inline src/input_common/sdl/sdl_impl.cpp:685]: ditto about the threshold default value

**Review by bunnei — APPROVED** [review]

**Pull request merged**